command/genesis/polybft_params.go

```diff
@@ -0,0 +1,224 @@
+)
+
+const (
+	validatorPrefixPathFlag = "prefix"
+
+	validatorSetSizeFlag = "validator-set-size"
+	sprintSizeFlag       = "sprint-size"
+	blockTimeFlag        = "block-time"
+	validatorsFlag       = "polybft-validators"
+
+	defaultEpochSize        = uint64(10)
+	defaultSprintSize       = uint64(5)
+	defaultValidatorSetSize = 100
+	defaultBlockTime        = 2 * time.Second
+
+	bootnodePortStart = 30301
+)
+
+var (
+	// hard code address for the sidechain (this is only being used for testing)
+	ValidatorSetAddr         = ethgo.HexToAddress("0xBd770416a3345F91E4B34576cb804a576fa48EB1")
+	SidechainBridgeAddr      = ethgo.HexToAddress("0x5a443704dd4B594B382c22a083e2BD3090A6feF3")
+	sidechainERC20Addr       = ethgo.HexToAddress("0x47e9Fbef8C83A1714F1951F142132E6e90F5fa5D")
+	SidechainERC20BridgeAddr = ethgo.HexToAddress("0x8Be503bcdEd90ED42Eff31f56199399B2b0154CA")
+)
+
+func (p *genesisParams) getPolyBftConfig(validators []*polybft.Validator) (*polybft.PolyBFTConfig, error) {
+	smartContracts, err := deployContracts()
+	if err != nil {
+		return nil, err
+	}
+
+	config := &polybft.PolyBFTConfig{
+		// TODO: Bridge
+		InitialValidatorSet: validators,
+		BlockTime:           p.blockTime,
+		EpochSize:           p.epochSize,
+		SprintSize:          p.sprintSize,
+		ValidatorSetSize:    p.validatorSetSize,
+		ValidatorSetAddr:    types.Address(ValidatorSetAddr),
+		SidechainBridgeAddr: types.Address(SidechainBridgeAddr),
+		SmartContracts:      smartContracts,
+	}
+
+	return config, nil
+}
+
+func (p *genesisParams) GetChainConfig() (*chain.Chain, error) {
+	validatorsInfo, err := ReadValidatorsByRegexp(path.Dir(p.genesisPath), p.validatorPrefixPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Predeploy staking smart contracts
+	genesisValidators := p.getGenesisValidators(validatorsInfo)
+
+	polyBftConfig, err := p.getPolyBftConfig(genesisValidators)
+	if err != nil {
+		return nil, err
+	}
+
+	extra := polybft.Extra{Validators: GetInitialValidatorsDelta(validatorsInfo)}
+
+	chainConfig := &chain.Chain{
+		Name: p.name,
+		Genesis: &chain.Genesis{
+			GasLimit:   p.blockGasLimit,
+			Difficulty: 0,
+			Alloc:      map[types.Address]*chain.GenesisAccount{},
+			ExtraData:  append(make([]byte, 32), extra.MarshalRLPTo(nil)...),
+			GasUsed:    command.DefaultGenesisGasUsed,
+			Mixhash:    polybft.PolyMixDigest,
+		},
+		Params: &chain.Params{
+			ChainID: int(p.chainID),
+			Forks:   chain.AllForksEnabled,
+			Engine: map[string]interface{}{
+				string(server.PolyBFTConsensus): polyBftConfig,
+			},
+		},
+		Bootnodes: p.bootnodes,
+	}
+
+	// set generic validators as bootnodes if needed
+	if len(p.bootnodes) == 0 {
+		for i, validator := range validatorsInfo {
+			// /ip4/127.0.0.1/tcp/10001/p2p/16Uiu2HAm9r5oP8Dmfsqbp1w2LdPU4YSFggKvwEmT6aTpWU8c8R13
+			bnode := fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", "127.0.0.1", bootnodePortStart+i, validator.NodeID)
+			chainConfig.Bootnodes = append(chainConfig.Bootnodes, bnode)
+		}
+	}
```

Comment:
I don't understand the need for this section.

Previously, you've turned off the bootnode flag as a required one, but here you enforce that all validators are bootnodes if nothing is provided.

cc @Kourin1996

Reply:
We want possibility to specify boot-nodes by parameter but, if not  all generated accounts (validators) will be boot nodes. This is functionality we used for our e2e tests.

Reply:
This was a helpful feature of `v3` that we wanted to port to `edge`. Having all the validators as `bootnodes` reduces the setup for local devnets.